Optimize throttle-based dynamic filter cutoff updates

etracer65 opened the pull request:
Fixes #7003 

Current logic produces excessive load because it updates the filter cutoffs every PID loop based on the throttle value.

The throttle values only change based on receive RX packets so they change relatively infrequently compared to the PID loop. However with rc smoothing on throttle the value can change every PID loop.  But in reality we don't really need to adjust the filter cutoffs for every tiny change to the floating point throttle value.

This change quantizes the throttle into 100 steps and uses that to compare to the previous value to decide if the filter cutoffs need to be updated.  While this reduces the resolution of the filter cutoffs it in turn dramatically reduces the processing overhead.  If needed the quantization steps can be increased for more resolution but at the cost of some performance.

butterSnake commented:
@etracer65 is there a debug method that can be used to compare between this and the initial version?

DieHertz commented on src/main/flight/mixer.c:
```diff
@@ -816,8 +821,14 @@ FAST_CODE_NOINLINE void mixTable(timeUs_t currentTimeUs, uint8_t vbatPidCompensa
+        dynLpfDTermUpdate(dynLpfThrottle);
+        dynLpfPreviousQuantizedThrottle = quantizedThrottle;
+    }
 #endif
```
`#undef DYN_LPF_THROTTLE_STEPS`

etracer65 replied:
I didn't actually define it here to imply scope to the function. Rather I was implying a dependence on `USE_DYN_LPF`. Since I already needed a block to declare the variable I included it here. I can more to the top.

DieHertz commented on src/main/flight/mixer.c:
```diff
@@ -747,6 +747,11 @@ void applyMotorStop(void)
 FAST_CODE_NOINLINE void mixTable(timeUs_t currentTimeUs, uint8_t vbatPidCompensation)
 {
+#ifdef USE_DYN_LPF
+#define DYN_LPF_THROTTLE_STEPS 100
```
Since preprocessor macros have no scope, I would either place this one somewhere at the top of this file, or undefine it after the last usage.

DieHertz commented on src/main/flight/mixer.c:
```diff
@@ -816,8 +821,14 @@ FAST_CODE_NOINLINE void mixTable(timeUs_t currentTimeUs, uint8_t vbatPidCompensa
+    if (quantizedThrottle != dynLpfPreviousQuantizedThrottle) {
+        // scale the quantized value back to the throttle range so the filter cutoff steps are repeatable
+        const float dynLpfThrottle = (float)quantizedThrottle / DYN_LPF_THROTTLE_STEPS;
+        dynLpfGyroUpdate(dynLpfThrottle);
```
No need to introduce loss of precision here, just do it like this:
```cpp
dynLpfGyroUpdate(throttle);
dynLpfDTermUpdate(throttle);
```

getting rid of `dynLpfThrottle` altogether.

ledvinap replied:
@DieHertz : This will actually increase error - throttle will be latched when entering given interval.
It will be indeterministic too - latched throttle position is different every time

DieHertz replied:
That's variability and I don't see how it's a bad thing. Just look at at how the same approach is used e.g. for grid-based motion planning, where each cell (bucket) is associated with continuous state, and a state mapped to a given cell can assume any values in given range, not just the edge values.

ledvinap replied:
PR code uses rounding - center of interval is used..

etracer65 commented:
@butterSnake No, not really.

etracer65 commented:
We've been discussing adding a minimum time interval on filter cutoff changes to prevent large amounts of throttle activity again tending back to excessive CPU load.  Additionally it's clear that updating the filter cutoff frequencies too often is detrimental to the filter performance as well.

So the question is what should the minimum update interval be? 1ms? 5ms? Something else?

@ledvinap @DieHertz Any thoughts on this? Are there some guidelines on how long we should let a filter "settle" on a newly updated cutoff frequency?  Maybe there's an ideal formula based on the actual cutoff?  It's possibly complicated by the fact that each of the 4 filters (gyro 1/2, dterm 1/2) can be either BIQUAD or PT1.

ledvinap commented:
@etracer65 : Caching throttle value should be good enough. Implementing minimum interval will complicate code, possibly to solve nonexistent problem.

PT1 just updates `alpha`, there should no detrimental effect after update.
Biquad section update should be OK too, but DF1 (suboptimal) implementation must be used

mikeller commented:
@ledvinap: I am not sure if possibly not solving an existing problem is the right approach - especially as this will be a hard to diagnose problem leading to unexpected flight behaviour.

ledvinap commented:
dynamic filter may be updated on RC rate, using unsmoothed throttle. This will ensure low update frequency and will change the filter without delay (no need to wait for ramp)

fujin commented:
Flew a branch including this included for multiple hours today @ 32/32/32 with dyn lpf gyro/dterm, I have blackbox fft_freq logging as well. I'll add a zip. Do we have a way forward or proposed changes from the review so far?

fujin commented:
Logs from test session today. If I can re-do these with any particular motions that are useful for establishing characteristics please let me know.

[LOGS.zip](https://github.com/betaflight/betaflight/files/2533798/LOGS.zip)

mikeller commented:
I'd still prefer if we added an upper bound for the filter update frequency - the result of the discussion seems to be that we can't be sure that we can't exclude the possibility that the current quantisation won't allow overloading.

etracer65 commented:
Add logic to prevent filter cutoff updates more often than every 5ms. Also moved the logic into its own function to tidy-up `mixTable()`.

While testing I added some blackbox logging (not included in commit) and confirmed the updates are behaving properly and at a minimum interval of 5ms.

ledvinap commented:
@etracer65 : My feeling is that quantization AND limiting update rate is overkill. Choosing one of them (probably constant update rate) will be cleaner.
And using constant update rate will be more deterministic, possibly avoiding hard to find bugs (triggered only by fast throttle changes)

mikeller commented on src/main/flight/mixer.c:
```diff
@@ -745,6 +751,27 @@ void applyMotorStop(void)
+    static timeUs_t lastDynLpfUpdateUs = 0;
+    static int dynLpfPreviousQuantizedThrottle = -1;  // to allow an initial zero throttle to set the filter cutoff
+
+
```
Extra newline.

mikeller commented on src/main/flight/mixer.c:
```diff
@@ -69,6 +69,12 @@ PG_REGISTER_WITH_RESET_TEMPLATE(mixerConfig_t, mixerConfig, PG_MIXER_CONFIG, 0);
 #define TARGET_DEFAULT_MIXER    MIXER_QUADX
 #endif
+
+#ifdef USE_DYN_LPF
```
Not sure if it's necessary to make the defines conditional - their names make clashes and use outside of the intended scope unlikely.
